common/http-api-client/src/lib.rs

```diff
@@ -215,95 +397,120 @@ impl Client {
     where
         B: Serialize + ?Sized,
         K: AsRef<str>,
-        V: AsRef<str>,
-    {
-        let url = sanitize_url(&self.base_url, path, params);
-        let mut request = self.reqwest_client.request(method.clone(), url);
-
-        if let Some(body) = json_body {
-            request = request.json(body);
-        }
-
-        request
-    }
-
-    pub fn create_get_request<K, V>(
-        &self,
-        path: PathSegments<'_>,
-        params: Params<'_, K, V>,
-    ) -> RequestBuilder
-    where
-        K: AsRef<str>,
-        V: AsRef<str>,
-    {
-        let url = sanitize_url(&self.base_url, path, params);
-        self.reqwest_client.get(url)
-    }
-
-    pub fn create_post_request<B, K, V>(
+        V: AsRef<str>;
+
+    /// Create an HTTP request using the host configured in this client and an API endpoint (i.e.
+    /// `"/api/v1/mixnodes?since=12345"`). If the provided endpoint fails to parse as path (and
+    /// optionally query parameters).
+    ///
+    /// Endpoint Examples
+    /// - `"/api/v1/mixnodes?since=12345"`
+    /// - `"/api/v1/mixnodes"`
+    /// - `"/api/v1/mixnodes/img.png"`
+    /// - `"/api/v1/mixnodes/img.png?since=12345"`
+    /// - `"/"`
+    /// - `"/?since=12345"`
+    /// - `""`
+    /// - `"?since=12345"`
+    ///
+    /// for more information about URL percent encodings see [`url::Url::set_path()`]
+    fn create_request_endpoint<B, S>(
         &self,
-        path: PathSegments<'_>,
-        params: Params<'_, K, V>,
-        json_body: &B,
+        method: reqwest::Method,
+        endpoint: S,
+        json_body: Option<&B>,
     ) -> RequestBuilder
     where
         B: Serialize + ?Sized,
-        K: AsRef<str>,
-        V: AsRef<str>,
+        S: AsRef<str>,
     {
-        let url = sanitize_url(&self.base_url, path, params);
-        self.reqwest_client.post(url).json(json_body)
+        // Use a stand-in url to extract the path and queries from the provided endpoint string
+        // which could potentially fail.
+        //
+        // This parse cannot fail
+        let mut standin_url: Url = "http://example.com".parse().unwrap();
```

Comment:
why can't we use the provided `endpoint` instead? (if we have to change the generic parameters slightly, I think that's fine). the standin here is quite odd to be honest

Reply:
The `url` crate wont parse things as a `URL` if they are missing a base or host. The stand-in url is there so that those pieces are guaranteed to be present. In this function the endpoint param is ONLY path and/or query parameters that we will add to the `base_url` for this Client in `create_request`. However because I wanted a default implementation for this there is no `base_url` available -- hence the stand-in.